### PR TITLE
storing two-way rules on equivalence db

### PR DESCRIPTION
**Pull request by christianbean**

### CODE REVIEW

File: comb_spec_searcher/strategies/constructor.py [hunk 1]
```diff
@@ -422,6 +430,235 @@ def __str__(self) -> str:
...
+            )
+        )
+        return tuple(sum(vals) for vals in zip(*mapped_params))
+
+    def _a(
+        self,
+        n: int,
+        parent_subterm: Callable[[int], Terms],
+        children_subterms: SubTerms,
+    ) -> Terms:
```
Comment (enadeau) [inline comb_spec_searcher/strategies/constructor.py:536]: Need to add some docstring about what are the a,b and c dict. I'll add a small explaination of how it works in the class docstring I think.

File: comb_spec_searcher/strategies/constructor.py [hunk 2]
```diff
@@ -422,6 +430,235 @@ def __str__(self) -> str:
...
+            ):
+                new_param = self._new_param(*(p for p, _ in param_value_pairs))
+                res[new_param] -= utils.prod((v for _, v in param_value_pairs))
+                assert res[new_param] >= 0
+        return res
+
+    def _c(
+        self,
+        children_subterms: SubTerms,
+    ) -> Terms:
```
Comment (enadeau) [inline comb_spec_searcher/strategies/constructor.py:556]: c should be cached

File: comb_spec_searcher/strategies/rule.py [hunk 3]
```diff
@@ -632,55 +651,64 @@ def join(res, new):
...
+            return "reverse of '{}'".format(self.strategy.formal_step())
+        return f"reverse of '{self.strategy.formal_step()}', counting child {self.idx}"
 
     def backward_map(
         self, objs: Tuple[Optional[CombinatorialObjectType], ...]
     ) -> Iterator[CombinatorialObjectType]:
-        yield cast(CombinatorialObjectType, self.original_rule.forward_map(objs[0])[0])
+        # We expect the object to be in only the first and in particular when
+        # mapped should be in the idx child.
+        assert isinstance(self.constructor, DisjointUnion)
```
Comment (enadeau) [inline comb_spec_searcher/strategies/rule.py:695]: Make something more detailed

File: comb_spec_searcher/strategies/rule.py [hunk 4]
```diff
@@ -632,55 +651,64 @@ def join(res, new):
...
+            CombinatorialObjectType, self.original_rule.forward_map(objs[0])[self.idx]
+        )
 
     def forward_map(
         self, obj: CombinatorialObjectType
     ) -> Tuple[Optional[CombinatorialObjectType], ...]:
-        return (next(self.original_rule.backward_map((obj,))),)
+        # We map forward to the first child.
+        # It is assumed that obj is ONLY on this child
+        assert isinstance(self.constructor, DisjointUnion)
```
Comment (enadeau) [inline comb_spec_searcher/strategies/rule.py:705]: make something more detailed

File: comb_spec_searcher/strategies/strategy.py [hunk 5]
```diff
@@ -310,6 +327,20 @@ def constructor(
...
+        idx: int,
+        comb_class: CombinatorialClassType,
+        children: Optional[Tuple[CombinatorialClassType, ...]] = None,
+    ) -> Constructor:
+        """
+        This is where the details of the 'reliance profile' and 'counting'
+        functions are hidden.
+        """
+        if children is None:
+            children = self.decomposition_function(comb_class)
```
Comment (enadeau) [inline comb_spec_searcher/strategies/strategy.py:342]: Those two lines are useless

File: comb_spec_searcher/strategies/strategy.py [hunk 6]
```diff
@@ -563,6 +638,10 @@ def __call__(
     def can_be_equivalent() -> bool:
         return False
 
+    @staticmethod
+    def is_two_way(comb_class: CombinatorialClassType) -> bool:
+        return False
```
Comment (enadeau) [inline comb_spec_searcher/strategies/strategy.py:643]: Shouldn't that be true? Same for can_be_equivalent above?

**Review by enadeau — COMMENTED** [review]
- [ ] The `.DS_Store` file should be removed

I left a few comments will work on fixing most of them

File: comb_spec_searcher/strategies/strategy.py (hunk 6 above)
Comment (christianbean) [reply comb_spec_searcher/strategies/strategy.py:643]: Since you have no children, it doesn't really matter

Comment (enadeau) [reply comb_spec_searcher/strategies/strategy.py:641]: Oh got tricked by the GitHub folding. I thought it was part of the symmetry strategy

**Review by enadeau — APPROVED** [review]

**Pull request merged**